Release 0.9.0 Updates

hunterjm opened the pull request:
This prepares for 0.9.0 by changing clips switches to recordings and updating media browser to use the new VOD functionality.

dermotduffy commented on custom_components/frigate/media_source.py:
```diff
@@ -136,7 +136,7 @@ def get_identifier_type(cls) -> str:
     def get_frigate_server_path(self) -> str:
         """Get the equivalent Frigate server path."""
-        return "/".join((self.get_identifier_type(), self.name))
+        return f"vod/event/{self.name}/index.m3u8"
```
(Apologies I don't know about this new end-point)

Will pre-existing clips continue to be accessible via this new path? If not we should probably use a different identifier type.

hunterjm replied:
yes, this endpoint first checks if a video exists under `/media/frigate/clips/{camera}-{event_id}.mp4` and will return that. Otherwise it will generate the clip from the recording files.

Everything is currently fully backwards-compatible.

dermotduffy commented on custom_components/frigate/views.py:
```diff
@@ -216,13 +216,26 @@ def _create_path(self, path: str, **kwargs: Any) -> str | None:
         return bool(config_entry.options.get(CONF_NOTIFICATION_PROXY_ENABLE, True))
 
+class VodProxyView(ProxyView):
+    """A proxy for clips."""
```
Can we retire the old proxy clips view?

hunterjm replied:
Probably yes. Also the recordings proxy view.

dermotduffy commented on custom_components/frigate/switch.py:
```diff
@@ -41,7 +41,7 @@ async def async_setup_entry(
             [
                 FrigateSwitch(entry, frigate_config, camera, "detect"),
-                FrigateSwitch(entry, frigate_config, camera, "clips"),
+                FrigateSwitch(entry, frigate_config, camera, "recordings"),
```
This will cause the old clips switch to be orphaned, as this effectively changes the unique_id. It would be  cleaner to add some logic here to explicitly delete the old clips switch from the entity registry before adding the new recordings switch.

dermotduffy replied:
(Is the intention to retire the clips switch? Why? It's very useful!)

hunterjm replied:
@blakeblackshear hinted at this below, but `clips` as a concept will be going away. There will be `events` with associated snapshots/thumbnails, and `recordings` which are for the cameras. You will be able to configure retention settings for `recordings` which are based on either a static retention period, or set different retention for recording segments based on the existence of overlapping events. That configuration is identical to the `clips` configuration today.

The reasoning behind this is to simplify the architecture, as well as remove duplicate `mp4` files for different events that overlap the same recording files.  For example, a car pulling into the driveway and 3 people getting out of it can potentially have 4 different events associated with it which would cause 4 nearly identical clips to be created.  This will allow us to only retain the recordings that overlap those events, and save a lot of storage space in the process.

hunterjm replied:
Your assumptions are correct, and yes, we still need to clean up the old entities from the registry.

dermotduffy commented on custom_components/frigate/views.py:
```diff
@@ -216,13 +216,26 @@ def _create_path(self, path: str, **kwargs: Any) -> str | None:
         camera = path.split("/")[0]
         if path.endswith("clip.mp4"):
-            return f"clips/{camera}-{event_id}.mp4"
+            return f"api/events/{event_id}/clip.mp4"
```
Does this work today? If not, it will presumably break the integration clip playing for anyone using an older version of Frigate. In that case, it would be good to explicitly check the server version against the expected one and print  a warning. 

Example from a different integration: https://github.com/home-assistant/core/blob/dev/homeassistant/components/hyperion/__init__.py#L157

hunterjm replied:
Yes, we should add a version check in here.

dermotduffy commented on custom_components/frigate/views.py:
```diff
@@ -216,13 +216,26 @@ def _create_path(self, path: str, **kwargs: Any) -> str | None:
+    extra_urls = ["/api/frigate/vod/{path:.*}"]
+
+    name = "api:frigate:vod"
+    requires_auth = False
```
Can we keep auth, given the existing clips proxy uses auth?

hunterjm replied:
The way HLS works, Media Browser will send the `authSig` query string parameter on the initial `index.m3u8` request. but it then gets processed and additional reaquests will be made for `segment-n.ts`, which will not have the `authSig` parameter, and thus fail this proxy.

I could potentially use https://github.com/kaltura/nginx-secure-token-module to pass `authSig` down to the generated segments, but need to do additional testing to see if that is feasible.

hunterjm replied:
This has been resolved, although it's a little weird. Home Assistant's URL signing policy only signs a single URL. In the case of HLS, that is the manifest file. After HLS loads the manifest, it proxies each segment through as well. I made a change to pass that authSig down to the segments, but then the request path doesn't match. What I ended up doing is disabling the built-in auth, and only validate that the parent path of the signature matches for each of the segments instead of the full path. That way people still can't directly enter a segment URL.

dermotduffy commented on custom_components/frigate/media_source.py:
```diff
@@ -287,10 +287,6 @@ class RecordingIdentifier(Identifier):
         default=None, validator=[attr.validators.instance_of((str, type(None)))]
     )
 
-    recording_name: str | None = attr.ib(
```
This will cause all existing recording identifiers to break. In practice, I think preserving and using these identifiers (e.g. to cast a particular recording) is probably a niche/rare/nonexistent usage.

hunterjm replied:
It's been a while since I wrote the media source integration, but I'm pretty certain these identifiers are only used internally. Casting requires a call to play media on a resolved identifier. Since Home Assistant doesn't currently support deep linking into the media browser, there is no reason to save this particular identifier, so we should be in the clear here.

dermotduffy commented on custom_components/frigate/media_source.py:
```diff
@@ -346,19 +340,22 @@ def get_frigate_server_path(self) -> str:
             out_parts.append(str(val))
+
+        if self.camera:
+            out_parts.append("index.m3u8")
```
I think you can just add this to `in_parts` definition above? Something like:
in_parts = [
...
   "index.m3u8" if self.camera else None
]

dermotduffy reviewed (commented):
[Understand this is draft, take these comments for what they're worth]

Apologies in advance, as I am missing 100% of the context around the 0.9.0 release so some of my questions may seem out of place. If there's some reference material or brief explanation of what the new VOD functionality does, that would really help for the PR description for those not involved in it.

Bunch of comments added. You'll also need tests, formatting fixes, etc as per failed build. Make sure you have the pre-commit installed and that formatting will automatically run.

blakeblackshear commented:
For broader context, clips as a concept is completely going away and everything is being consolidated into recordings. The current version effectively records the video stream twice. Now, 24/7 recordings will be retained based on whether or not events overlap, and footage is dynamically available from the underlying recordings through the VOD feature of nginx which can assemble a playlist of mp4 files.

S-Przybylski commented:
Dear @blakeblackshear (I know that a pull request is not the right discussion point - please do not be mad at me),
please be aware that many people (like me) use a simple RASPI with a SD-card for storing the last 10-15 days activities. I personally use six cams with a dedicated raspi currently. 
If I got you right: A recording need to be set up using this future release in general. Then its practical no longer possible to store "important activities" on a SD card. Additional hardware (external SSD hardware + power supply for an additional hub will be required). Am I correct?

blakeblackshear commented:
That's not correct. Frigate will still only store the video necessary for "important events". 24/7 recording will not be required. Frigate will "record" video from your camera one time. You can configure it to save recordings 24/7 or just for where the contain footage of events.

S-Przybylski commented:
Thats really cool! Thanks for correcting me! 
Will it later be possible to select the default recording method like "24/7, motion, person, schedule, schedule+motion/person,...?

drinfernoo commented:
@blakeblackshear, might this "record once" method help with performance of cameras with somewhat unreliable RTSP streams, like the Wyze cams?

blakeblackshear commented:
> @blakeblackshear, might this "record once" method help with performance of cameras with somewhat unreliable RTSP streams, like the Wyze cams?

I don't see how it would.

drinfernoo commented:
@blakeblackshear Ah, well 😅 I guess I was thinking the new strategy was meant to reduce the number of streams coming from the camera at any given time... But no worries 😉

hunterjm commented:
> [Understand this is draft, take these comments for what they're worth]

Thanks for the thorough review. This is definitely still a WIP that is based off of https://github.com/blakeblackshear/frigate/pull/1352. It was a very rough first pass to get 90% of things working together. There will probably be modifications to both PRs incoming (which will include addressing your comments).

> That's not correct. Frigate will still only store the video necessary for "important events". 24/7 recording will not be required. Frigate will "record" video from your camera one time. You can configure it to save recordings 24/7 or just for footage of events.

This isn't entirely true. Those that are using Frigate with a file system on an SSD **will** see a significantly larger number of reads/writes to that SD card as previously the recording wasn't moved out of `tmpfs` unless it matched an event, and now recordings will be moved from `tmpfs`, but recordings found not to match an event will be deleted shortly thereafter.

The amount of space required does not change, but your SD card will see additional wear and will likely fail faster with the current implementation.

drinfernoo commented:
That's kind of not the best news 😬

hunterjm commented:
> That's kind of not the best news 😬

To be fair, I said “more wear”. This is an NVR project (emphasis on the R which stands for Recorder). Recording video is fundamentally write heavy. I wouldn’t recommend enabling even the current clips implementation if you are saving it to an SD card.

dermotduffy reviewed (commented):
Thanks for getting rid of the old proxies and figuring out an auth scheme!

Should we do a similar cleanup for the media identifiers given the cleanup on the backend? e.g. I could imagine something like:

ClipIdentifier -> Remove
ClipSearchIdentifier -> Rename to EventSearchIdentifier (or similar)
RecordingIdentifier -> Keep

hunterjm commented:
> Thanks for getting rid of the old proxies and figuring out an auth scheme!
> 
> Should we do a similar cleanup for the media identifiers given the cleanup on the backend? e.g. I could imagine something like:
> 
> ClipIdentifier -> Remove
> ClipSearchIdentifier -> Rename to EventSearchIdentifier (or similar)
> RecordingIdentifier -> Keep

Sounds reasonable. Still planning on combing back through this and adding tests before taking it out of draft status. I’ll pull it out of draft when I’m done. I typically just like to get things working and clean it up after.

I have this (along with the other PR) running in prod now and will let it sit for a few days to try and catch any edge cases.

one I’ve found already is the clip notification proxy endpoint doesn’t load the video properly in iOS Safari. Works fine in desktop Chrome. Will need to do further digging on what exactly is the issue before this is ready.

dermotduffy commented:
FYI: @hunterjm You'll need to do an unpleasant merge with https://github.com/blakeblackshear/frigate-hass-integration/pull/116 , but I did some of the renames to move ClipSearchIdentifier -> EventSearchIdentifier to prepare in a minor way for this PR when you're ready.

hunterjm commented:
> FYI: @hunterjm You'll need to do an unpleasant merge with #116 , but I did some of the renames to move ClipSearchIdentifier -> EventSearchIdentifier to prepare in a minor way for this PR when you're ready.

No worries. My free time lately has been practically non-existant, but I haven't forgotten about this. Please don't feel like you need to halt forward progress waiting for me. I am intimately familiar with resolving merge conflicts :)

blakeblackshear commented:
@dermotduffy I finally got everything to pass. I still need to update the config in the test fixture data. Do you have a good example of how I should go about deleting the old clips switch? Also, do you mind doing another review? I think there are enough changes here that we shouldn't bother trying to maintain backwards compatibility with older versions of frigate.

dermotduffy reviewed (commented):
Looking good, mostly small stuff. On the two larger topics:

*  Backwards compatibility: It doesn't seem _that_ incompatible. Is there already a server release that is backwards compatible? If yes, we could add a warning and release now to get people to upgrade in advance. But I'm assuming not, and both server/integration must be updated in lockstep. If that's the case, we should add a warning when the integration is loaded telling people to upgrade, to try to reduce the support load. See [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/hyperion/__init__.py#L156) for an example of where I did similar.
* On deleting the clips switch, I don't have a personal example to hand at the entity level, but it should be something like this, probably in `async_setup_entry` (totally untested pseudocode):

```
from homeassistant.helpers import entity_registry as er
[...]
unique_id = get_frigate_entity_unique_id(config_entry_id, SWITCH_DOMAIN, "clips")
entity_registry = er.async_get(hass)
entity_id = entity_registry.async_get_entity_id(SWITCH_DOMAIN, DOMAIN, unique_id)
if entity_id:
  entity_registry.async_remove(entity_id)
```

[Here's an example I found](https://github.com/home-assistant/core/blob/dev/homeassistant/components/tasmota/discovery.py#L72) from someone else that looks similar.
 
PS: If you haven't, recommend installing the pre-commit in your local repo if it isn't, helps spots the lint/typing issues without needing to wait for the github build  to slowly churn.

blakeblackshear commented on custom_components/frigate/views.py:
```diff
@@ -216,14 +196,92 @@ def _create_path(self, path: str, **kwargs: Any) -> str | None:
+            _LOGGER.debug("Missing refresh token.")
+            return False
+
+        request[KEY_HASS_USER] = refresh_token.user
```
@hunterjm I commented these out and everything seems to work fine. Any insight?

blakeblackshear commented on custom_components/frigate/media_source.py:
```diff
@@ -1277,17 +1274,10 @@ def _browse_recordings(
 
         for recording in recordings:
             title = self._generate_recording_title(identifier, recording)
-            if not title:
```
I haven't been able to figure out how to actually get a ValueError to result in the title being None in this context because all the parts of the identifier are validated upstream when parsing the identifier. Also, the folder contents under the camera are longer parsed since recordings are now hour long VOD playlists.

dermotduffy replied:
I think this can still happen with a syntactically valid but semantically invalid date, i.e. [this test](https://github.com/blakeblackshear/frigate-hass-integration/blob/master/tests/test_media_source.py#L848) I think is triggering the earlier catch of the ValueError. The upstream validation is checking they are valid numbers, not necessarily valid dates when assembled.

If I'm missing something and this is really just dead code, then lets just remove the exception handling of ValueError in `_generate_recording_title` rather than catching it but not handling the `None`. That'll cause the ValueError to propagate all the way to the top which I think is appropriate if we think this cannot ever happen.

blakeblackshear replied:
I can hit the ValueError in the `test_async_browse_media_recordings_root` test, but I can't find a way to hit it when it's being called from the `_browse_recordings` method where I removed this check. The exception handling isn't dead code, but it doesn't seem possible to have title be None in this particular method. I couldn't get back to 100% coverage without removing this check.

blakeblackshear commented:
Thanks for the review. I think I have addressed all the feedback with the exception of the check for a None title. I think that may just be dead code.

For backwards compatibility, I just don't think its worth it. There are no new features in this version, so no benefit to upgrade if you are not on a newer version of Frigate. I think I would prefer to output a warning in the logs. In addition to the change for clips, the recordings now stop at the camera level and let you play a 1 hour playlist instead of drilling down an additional level to view the files under a camera which are now 10s segments rather than 1 minute segments. Seems like enough if/else logic to avoid bloating the code with.

hunterjm commented:
Thanks for picking this up Blake, and sorry for the hit and run here with these PRs. Life has been a little crazy. I agree that backwards compatibility is not necessary. I would probably do a major version bump on the HASS integration and include a Version check in the code. I wouldn't print out a warning, I would actually error out of the integration setup if the server version isn't >= 0.9.0. Then the major version bump for the HACs release **should** let people know there are breaking changes.

dermotduffy commented on custom_components/frigate/__init__.py:
```diff
@@ -152,6 +165,19 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        unique_id = get_frigate_entity_unique_id(
+            entry.entry_id, SWITCH_DOMAIN, f"{camera}_clips"
+        )
+        entity_registry = er.async_get(hass)
```
Nit: You can put this outside the `for` loop.

dermotduffy commented on custom_components/frigate/__init__.py:
```diff
@@ -141,6 +144,16 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     except FrigateApiClientError as exc:
         raise ConfigEntryNotReady from exc
 
+    if server_version < FRIGATE_VERSION_ERROR_CUTOFF:
```
Nit: That's going to end up going a lexical comparison of the strings rather than a version comparison. That's probably okay here, but better off wrapping it in `AwesomeVersion` for best practice ([example](https://github.com/home-assistant/core/blob/dev/homeassistant/components/hyperion/__init__.py#L156)).

dermotduffy commented on custom_components/frigate/views.py:
```diff
@@ -250,39 +246,26 @@ async def _async_validate_signed_manifest(self, request: web.Request) -> bool:
 
         if signature is None:
-            _LOGGER.debug("Missing signature.")
+            _LOGGER.error("Missing authSig query parameter on VOD segment request.")
```
Nit: `warning` probably, as the integration itself won't have an error condition as a result. Same below.

dermotduffy commented on tests/__init__.py:
```diff
@@ -288,7 +288,7 @@ async def setup_mock_frigate_config_entry(
     config_entry: ConfigEntry | None = None,
     client: AsyncMock | None = None,
 ) -> ConfigEntry:
-    """Add a mock MotionEye config entry to hass."""
```
Oops! (Thanks)

dermotduffy commented on tests/test_init.py:
```diff
@@ -158,6 +169,65 @@ async def test_entry_migration_v1_to_v2(hass: HomeAssistant) -> None:
+        )
+        is not None
+    )
+
```
Maybe zip through the loop of the other entities and make sure they all still exist except the clips switch?

dermotduffy reviewed (commented):
Thanks for the fixes!

(Personally, I think the warning you've added is better than failing the integration load entirely when we think most it will still work correctly).

blakeblackshear commented:
@dermotduffy given that I'm not that familiar with how HACS version and releases work, how do you recommend we proceed to get a pre-release version out?

dermotduffy commented:
@blakeblackshear How far away timewise do you think the official release of 0.9.0 is?

Once we do a pre-release here that'll start steering people towards 0.9.0 fairly promptly I'd guess, as opting in to betas is a front-and-center option in HACS on every initial install. So I wouldn't recommend it until you're confident, there's an official 0.9.0 almost ready and installation instructions clear etc, e.g. within O(1 week) of the official release? Are we at that point?

We should also write a few words about the HA migration (e.g. clips switch is gone). Is there a place for such documentation regarding what's new in 0.9.0? We can just add it to the README.md here until this is all behind us, but if you have a central 0.9.0 place we could add something there instead.

Other than that, there's no particular magic with HACS releases. We'll just cut a github pre-release, cover the release notes in alarm bells / flame icons, and update support documentation as the requests arrive.

blakeblackshear commented:
I plan to publish a release candidate this week and update the beta addon as well. When that happens, we should cut the release here. I typically document changes for new releases on the release in the frigate repo.

dermotduffy commented:
Sounds good. If your release notes will include impact on HA, we can just link to them from the integration/HACS release notes.

dermotduffy commented:
@blakeblackshear Any reason we shouldn't merge this now? (If we'll be cutting a pre-release in a few days, might be good to get it merged now for people who want to test  before that point?)
